### PR TITLE
Don't reload play application if only assets change

### LINKED ISSUE #3295: When reloading a Play! application, if the classloader has not changed, do not reload the application.
When changing a file that is only served as "assets" (and is not part of the classloader), Gradle still reloads the Play application. This is unnecessary.

SBT had a related problem: https://github.com/playframework/playframework/issues/2905

### Expected Behavior

When changing an asset file or a file that does not affect the classloader of the Play application, the Play application does not reload.  The Gradle build may still need to run (e.g., to copy assets/generate other assets).

### Current Behavior

Play application reloads on any change.

### Steps to Reproduce (for bugs)

1. Using the "basic" Play application sample, run the Play application (`gradle runPlay`)
2. Change a file in `public/`.  Observe the Play application stops/starts.  This is unnecessary when the file is not a part of the classloader.

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

This fixes #3295 . Previously, Play server will reload when any files change (controllers/views/assets, etc.). However, it's not necessary to reload the server since it reads assets from assets directory directly. This PR puts a monitor which watches non-assets files change in `PlayApplicationRunner`, and tells Play server not to reload in this situation.

### CODE REVIEW

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java [hunk 1]
```diff
@@ -16,35 +16,102 @@
...
         process.getConnection().connect();
         playApplication.waitForRunning();
         return playApplication;
     }
 
+    private class PlayClassloaderMonitorDeploymentDecorator implements Deployment {
+        private Deployment delegate;
+        private FileCollection applicationClasspath;
+        private HashCode snapshot;
+        private boolean isPlay22;
```
Comment (bigdaz) [inline subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java]: Mark non-mutable fields `final`.

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java [hunk 2]
```diff
@@ -16,35 +16,102 @@
...
+
+            if (!delegateStatus.hasChanged()) {
+                return delegateStatus;
+            }
+
+            HashCode oldSnapshot = updateSnapshot();
+
+            if (!snapshot.equals(oldSnapshot)) {
+                return delegateStatus;
+            } else {
```
Comment (bigdaz) [inline subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java]: For readability I'd extract an `applicationClasspathChanged()` method that hides the implementation details of snapshotting, and change this to:
```
if (!applicationClasspathChanged()) {
    return delegateStatus
}
return new Status() {
   ...
```

File: subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/AbstractMultiVersionPlayReloadIntegrationTest.groovy [hunk 3]
```diff
@@ -18,7 +18,23 @@ package org.gradle.play.integtest.fixtures
...
 @TargetCoverage({ JavaVersion.current().isJava8Compatible() ? PlayCoverage.PLAY23_OR_LATER : PlayCoverage.PLAY23_OR_EARLIER })
 abstract class AbstractMultiVersionPlayReloadIntegrationTest extends AbstractMultiVersionPlayContinuousBuildIntegrationTest {
+    protected serverRestart() {
+        ConcurrentTestUtil.poll {
+            assert serverStartCount() > 1
+        }
+    }
+
+    protected serverNotRestart() {
+        if (!versionNumber.toString().startsWith("2.2")) {
```
Comment (bigdaz) [inline subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/AbstractMultiVersionPlayReloadIntegrationTest.groovy]: `serverNotRestart` doesn't sound right. I think `noServerRestart` reads better as an assertion.

File: subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/AbstractMultiVersionPlayReloadIntegrationTest.groovy [hunk 4]
```diff
@@ -18,7 +18,23 @@ package org.gradle.play.integtest.fixtures
...
+            assert serverStartCount() == 1
+        }
+    }
+
+    protected serverStartCount() {
```
Comment (bigdaz) [inline subprojects/platform-play/src/testFixtures/groovy/org/gradle/play/integtest/fixtures/AbstractMultiVersionPlayReloadIntegrationTest.groovy]: If you make this `getServerStartCount()` then you can use `assert serverStartCount == 1`.

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java (hunk 2 above)
Comment (blindpirate) [reply subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayApplicationRunner.java]: Good catch, that's much neater.

**Comment by blindpirate** [comment]
Thanks @bigdaz !

**Pull request merged**